### PR TITLE
driver orsay: improve imaging reliability and more

### PR DESCRIPTION
**Pull request by pieleric**
This should mostly complete the support for imaging on the Orsay Physics FIB.
It also fixes a few other issues related to reliability found while testing the image acquisition. 
Now the Odemis GUI support playing the image, adjusting the acquisition setting, adjusting the focus.

There is still one bug in the FIB software which is not handle when changing the RoI up in Y. Still hoping that Orsay Physics can directly fix this issue instead of writing a workaround.

For the reviewers: I recommend checking commit per commit, as it might clarify the changes.

### CODE REVIEW

File: src/odemis/driver/test/orsay_test.py [hunk 1]
```diff
@@ -2026,132 +2080,265 @@ def setUpClass(cls):
...
+        finally:
+            self.detector.data.unsubscribe(self.receive_image)
+
+        logging.info("Durations: %s", durations)
+
+    # TODO: test cropping (while acquiring)
 
     def test_get_varying_dwell_time(self):
-        self.skipTest(f"By default this test is skipped. It test the functionality on changing dwell times. User input is required")
+        # self.skipTest(f"By default this test is skipped. It test the functionality on changing dwell times. User input is required")
```
Comment (Kleijwegt) [inline src/odemis/driver/test/orsay_test.py:2291]: Probably this comment can go away? (or if changing dwelltimes is not that stable yet it should remain uncommented?)

File: src/odemis/driver/orsay.py [hunk 2]
```diff
@@ -1547,41 +1591,42 @@ def disconnect(self):
...
 
         :param (Orsay Parameter) parameter: The parameter on the Orsay server that calls this callback
-        :param (str) attr_name: The name of the attribute of parameter which was changed
+        :param (str) attr_name: The name of the attribute of parameter which was changed (typically, "Actual")
         """
-        # Check that the value of the attr_name argument makes sense
-        if attr_name is None:
-            attr_name = self._attr_name
-        if attr_name != self._attr_name:
+        # Drop events when something else than the "Actual" attribute changed (eg, Target).
```
Comment (Kleijwegt) [inline src/odemis/driver/orsay.py:1606]: Spelling mistake? (**e.g.** Target)

File: src/odemis/driver/orsay.py [hunk 3]
```diff
@@ -2250,11 +2310,19 @@ def __init__(self, name, role, parent, **kwargs):
...
+        self.imageFormat = model.ResolutionVA(SCANNER_MAX_RES, rng=(SCANNER_MIN_RES, SCANNER_MAX_RES),
                                               setter=self._imageFormat_setter)
-        self.resolution = model.TupleContinuous((1024, 1024), unit="px", range=[(1, 1), (1024, 1024)],
+        self.resolution = model.TupleContinuous(SCANNER_MAX_RES, unit="px", range=[(1, 1), SCANNER_MAX_RES],
                                                 setter=self._resolution_setter)
-        self.translation = model.TupleContinuous((0.0, 0.0), unit="px", range=[(-512.0, -512.0), (512.0, 512.0)],
+
+        # It uses the pixel convention: smallest value corresponds to top left of the image.
+        # The unit is the pixel at scale == 1.
+        # We accept floats, because it's Odemis convention, however, it will always be rounded
```
Comment (Kleijwegt) [inline src/odemis/driver/orsay.py:2322]: I see you use "round" to round the float. I think it is nice to add that info also here so that a reader knows using which function the float is rounded to an int.

File: src/odemis/driver/orsay.py [hunk 4]
```diff
@@ -1316,8 +1315,8 @@ def _updateTemperature(self, parameter=None, attr_name="Actual"):
         """
         if parameter is None:
             parameter = self._temperaturePar
-        if float(self._temperaturePar.Actual) == float(self._temperaturePar.Target):
-            logging.debug("Target temperature reached.")
+        # if float(self._temperaturePar.Actual) == float(self._temperaturePar.Target):
+        #     logging.debug("Target temperature reached.")
```
Comment (tepals) [inline src/odemis/driver/orsay.py:1319]: why is this commented?

**Review by tepals — CHANGES_REQUESTED** [review]

File: src/odemis/driver/test/orsay_test.py (hunk 1 above)
Comment (pieleric) [reply src/odemis/driver/test/orsay_test.py:2291]: Hi Kornee, nice to have you shinning up here!
The change of dwell time is now stable (as far as I could test). So indeed this line can completely go away.

File: src/odemis/driver/orsay.py (hunk 4 above)
Comment (pieleric) [reply src/odemis/driver/orsay.py:1319]: Too many logs. It's still handy to know how to check that the target temp is reached, so I left the code.

File: src/odemis/driver/orsay.py [hunk 5]
```diff
@@ -2471,51 +2475,50 @@ def _updateImageFormat(self, parameter=None, attr_name="Actual"):
...
+
+        if new_value[0] != new_value[1]:
+            # For the scale, we only support square image format
+            logging.warning("Image format received is not square: %s", new_value)
+
+        if self.imageFormat._value != new_value:
+            self.imageFormat._value = new_value  # to not call the setter
+            self.imageFormat.notify(new_value)
 
     def _clip_and_set_image_area(self, target_resolution, target_translation):
```
[the remark this answers is not in the document]
Comment (pieleric) [reply src/odemis/driver/orsay.py:2531]: I'd rather not, because it's a private function. I prefer doing just "black box" testing, by only using the standard interface. So it's indirectly tested when testing the .translation and .resolution VAs, but not explicitly tested.

File: src/odemis/driver/orsay.py [hunk 6]
```diff
@@ -2584,21 +2585,25 @@ def _updateTranslationResolution(self, parameter=None, attr_name="Actual"):
...
+        new_translation = (area[0] - self.imageFormat.value[0] / 2 + area[2] / 2,
+                           area[1] - self.imageFormat.value[1] / 2 + area[3] / 2)
 
         new_resolution = tuple(area[2:4])
 
-        self.translation._value = new_translation  # to not call the setter
-        self.resolution._value = new_resolution  # to not call the setter
+        if (self.translation._value != new_translation or
+            self.resolution._value != new_resolution
+        ):
```
[the remark this answers is not in the document]
Comment (pieleric) [reply src/odemis/driver/orsay.py:2639]: Yeah... the problem with `if (` is that it's exactly the size of one indentation... so pretty much any formatting looks odd. I've added a line after the opening bracket.

File: src/odemis/driver/orsay.py [hunk 7]
```diff
@@ -2891,133 +2915,214 @@ def __init__(self, name, role, parent, **kwargs):
...
+        scan_time = self._scanner.dwellTime.value * len(img)
+        metadata[model.MD_ACQ_DATE] = now - scan_time
+
+        # Crop to the current area
+        img_area = list(map(int, img_area.split(" ")))
+        img_size = list(map(int, img_size.split(" ")))
+        if len(img_area) != 4 or len(img_size) != 2:
+            logging.error("Cannot crop, img_area decoded to %s, img_size to %s", img_area, img_size)
+        elif not (0 <= img_area[0] < img.shape[1] and
+                  0 <= img_area[1] < img.shape[0] and
```
[the remark this answers is not in the document]
Comment (pieleric) [reply src/odemis/driver/orsay.py:3030]: Yes, because number has Y as first dim and X as second dim. I'll add a comment.

File: src/odemis/driver/test/orsay_test.py [hunk 8]
```diff
@@ -2026,132 +2080,265 @@ def setUpClass(cls):
             raise unittest.SkipTest(NO_SERVER_MSG)
 
         cls.oserver = orsay.OrsayComponent(**CONFIG_ORSAY)
-        cls.datamodel = cls.oserver.datamodel
         for child in cls.oserver.children.value:
-            if child.name == CONFIG_FIBBEAM["name"]:
-                cls.fibbeam = child
-            elif child.name == CONFIG_SCANNER["name"]:
+            # if child.name == CONFIG_FIBBEAM["name"]:
+            #     cls.fibbeam = child
```
[the remark this answers is not in the document]
Comment (pieleric) [reply src/odemis/driver/test/orsay_test.py:2085]: yes, unused... but of course it can now be entirely dropped.

File: src/odemis/driver/test/orsay_test.py [hunk 9]
```diff
@@ -2026,132 +2080,265 @@ def setUpClass(cls):
...
-        self.assertEqual(image.shape, self.fibbeam.imageFormat.value)
+        self.assertEqual(image.shape, self.scanner.resolution.value)
+
+    def test_scale(self):
+        """
+        Check that the image is received at the specified scale
+        """
+        max_res = self.scanner.resolution.range[1]
+
+        res = max_res
```
[the remark this answers is not in the document]
Comment (pieleric) [reply src/odemis/driver/test/orsay_test.py:2126]: I'm now using it in more places ;-)

File: src/odemis/driver/test/orsay_test.py [hunk 10]
```diff
@@ -2026,132 +2080,265 @@ def setUpClass(cls):
...
+        self.scanner.scale.value = (1.0, 1.0)
+        self.scanner.resolution.value = res
+        self.assertEqual(self.scanner.scale.value, (1, 1))
+        self.assertEqual(self.scanner.resolution.value, res)
```
[the remark this answers is not in the document]
Comment (pieleric) [reply src/odemis/driver/test/orsay_test.py:2130]: I don't mind. For now that's more PyCharm than a general case, as in the unittest code, it's just called "first" and "second".

**Comment by pieleric** [comment]
I've updated the commits with fixes from the review. (Use the "Compare" button to see all the changes since the previous review).
I've also updated the code to work better for some corner cases with the RoI. Also, in the meantime, the new version of the Orsay Physics server has several issues resolves, which allow to run almost all test cases now. 
I've also added a new commit to adjust the beam shift so that it follows the odemis convention.

**Review by FCanberk — APPROVED** [review]

**Comment by FCanberk** [comment]
@tepals @nandishjpatel is it possible that this PR is overlooked? I kinda need this PR merged for my branch :smile:

**Review by tepals — APPROVED** [review]
one small comment
